### DISCUSSION TITLE
Achieving Multi-tenancy using Micronaut-data with MongoDB

### DISCUSSION THREAD

**Discussion by Saumya999** [micronaut-projects/micronaut-data#2541] [Q&A]
I want to achieve multi tenancy using MongoDB , I am using Micronaut Kotlin with Micronaut-data for database connectivity. 
I am not using HttpTenantResolver provided by Micronaut. I wrote my own logic in filter chain. 

**_Now I have the tenant-id depend ending on that with some logic I want to override the mongoDb connection String mentioned in my application.yml . can anyone please help me how I can do that ?_** 

and I avoided in built HttpTenantResolver because I don't know how many tenants will be there. It will dynamically increase - for each tenant database will be different with common collection schema ( with different data in it ). That's why it's impossible to maintain all the datasource connection string in Yml. It should be built dynamically in runtime. 

thanks in advance

**Comment by graemerocher** [comment]
does this not help? https://micronaut-projects.github.io/micronaut-data/latest/guide/#multitenancy

**Reply by Saumya999** [reply]
If you see in the Doc - they are maintaining 2 server urls in the application.yml for FOO and Bar. 
Here my requirement is little different. But in my case beforehand I don't know how many tenants will be there. 

For example: 
one endpoint = **/register** that will be public which doesn't need any tenantId in Header . It will create and registration id for the customer. Basically That generated Id will be tenant-id for respective calls. and tenant-Id will be the databse Name. Database will contain - per customer products, orders, billings etc as collections. (B2B) 

application.yml

```
mongodb:
  uri: 'mongodb://localhost:27017/{tenantId}
```

This is something I need. - For now I am mentioning one fixed value in place of tenantId variable. 
  
  this tenantId I am getting below 
  
  ```
@Singleton
@Filter("/**")
class TenantFilter(private val connectionStorage: ConnectionStorage) : HttpServerFilter{

    // TODO Multi-Tenancy implementation to add x-tenant-id in Header for dynamic database

    override fun doFilter(request: HttpRequest<*>?, chain: ServerFilterChain?): Publisher<MutableHttpResponse<*>> {
        val attribute: @Nullable String = request?.headers?.get("x-tenant-id") ?: throw RuntimeException("Missing Tenant Header")
        if (chain != null) {
            connectionStorage.setConnection(attribute)
            return chain.proceed(request)
        } else {
            throw RuntimeException("proper request chaining not happening")
        }
    }


}
```

```
@Singleton
class ConnectionStorage {
    private var threadLocal:ThreadLocal<String> = ThreadLocal()
    fun setConnection(tenantId: String) {
        threadLocal.set(tenantId)
    }
    fun removeConnection(tenantId: String) {
        threadLocal.remove()
    }
    fun getConnection(): String {
        return threadLocal.get()
    }
}
```
  
  
  Now tenantId is set in ThreadLocal storage. I need to dynamically override the connection String of MongoDB in runtime. 
  
  @graemerocher  Please check if you could help and save me :)

**Comment by dstepanov** [comment]
Please don't use thread-local but our new propagation API. (see Core docs)
The only way right now would be to replace the current `MongoClient` definition with some proxy that will delegate to the tenant client.

**Reply by Saumya999** [reply]
Can you give me some code examples how to achieve this ? I didn't get
